zk.el

```diff
@@ -309,6 +296,31 @@ otherwise just match against `zk-file-name-regexp'."
   "Substitute `zk-file-name-separator' with spaces in STRING."
   (replace-regexp-in-string zk-file-name-separator " " string))
 
+(defun zk-file-name-regexp ()
+  "Return the correct regexp matching zk file names.
+The regexp captures these groups:
+
+Group 1 is the zk ID.
+Group 2 is the title."
+  (concat "\\(?1:" zk-id-regexp "\\)"
+          "."
+          "\\(?2:.*?\\)"
+          "\\."
+          zk-file-extension
+          ".*"))
+
+(defun zk--file-name-id (file)
+  "Return the ID of the given FILE.
+This relies on match data from `zk-file-name-regexp`."
+  (match-string-no-properties 1 file))
```

Comment:
I don't think these instances of `match-string` should be extracted. Since they rely on match data, they are contingent functions and don't have any use independent of very specific conditions (ie, following an instance where there is the appropriate match data provided by `zk-file-name-regexp`.

Reply:
I see your point, but creating accessors like this allows to abstract the meaning of the code away from implementation. If you change what groups 1 and 2 mean, for example, then you will need to find all instances of `(match-string[-no-properties]` and update them. Having the accessors in place means that the code will need to be changed in one place only (i.e. accessors).

Also, the next step in refactoring might be to put all the string matching into the functions themselves, so the whole construct of
``` emacs-lisp
(when (match-string (zk-file-name-regexp) file)
  (replace-regexp-in-string zk-file-name-separator
                            " "
                            (match-string-no-properties 2 file))
```
would become just `(zk-file-name-title file)`. At that point, one could rewrite `zk--parse-file` to the following:

``` emacs-lisp
(defun zk--parse-file (target files)
  (let ((result
         (mapcar (pcase target
             ('id #'zk--file-name-id)
             ('title #'zk--file-name-title)
             (_
               (signal 'wrong-type-argument '(target (or 'id 'title)))))
            (if (listp files)
                files
              (list files)))))
    (if (cdr result)
        result
      (car result))))
```

In other words, introducing a layer of abstraction for raw data has benefits, I feel.

Reply:
Under what circumstances would the meaning of groups 1 and 2 be changed? Their meaning as ID and FILE is pretty much essential, right?

As they are currently, `zk--file-name-id` and `-title` are almost just aliases for `match-string`.

Perhaps they would be easier to justify if they did the string matching themselves, but that would mean that each function would run `string-match` twice (once for id and once for title) instead of just once. So, it's kind of splitting hairs, no?

Reply:
True, it's hard to imagine them changing. So, yes, they are definitely just aliases for `match-string`. I suppose at the core of it not liking "magical numbers," and `(zk--file-name-title file)` is more readable than `(match-string 1 file)`. If only Emacs regular expression syntax had named groups like other modern languages rather than just merely numbered ones (so one could write regexps like `\\(?<id>[0-9]\\{12\\}\\).\\(?<title>.*\\)`). Maybe one day.

I get your point about doing `string-match` twice. The performance should not be affected much, but it's your package, so if you prefer to keep that part as is, I can `git revert` the relevant commits, just keep the `zk-file-name-regexp` part.

Reply:
I was curious about performance, so after moving `string-match` into the `zk--file-name-id` and `zk--file-name-title`, I instrumented all `zk-` functions with `elp.el` and spent half an hour or so "conversing" with my Zettelkasten. The results are not bad, but the added total of three seconds (compared to 8 seconds for `zk-index-search') is probably not worth it, indeed.

[![zk--file-name-title and zk--file-name-id take up 3 seconds for 100K calls each, zk-file-name-regexp just a half-second despite 350K calls.](https://i.postimg.cc/3w3XBksJ/2022-10-18-ELP-Results.jpg)](https://postimg.cc/xXFNj1jW "zk--file-name-title and zk--file-name-id take up 3 seconds for 100K calls each, zk-file-name-regexp just a half-second despite 350K calls.")